test/1_stdlib/Collection.swift

```diff
@@ -202,5 +202,41 @@ func testIsEmptyFirstLast() {
 }
 testIsEmptyFirstLast()
 
+/// A `CollectionType` that vends just the default implementations for
+/// `CollectionType` methods.
+struct CollectionOnly<T: CollectionType> : CollectionType {
+  var base: T
+
+  var startIndex: T.Index {
+    return base.startIndex
+  }
+
+  var endIndex: T.Index {
+    return base.endIndex
+  }
+
+  func generate() -> T.Generator {
+    return base.generate()
+  }
+
+  subscript(position: T.Index) -> T.Generator.Element {
+    return base[position]
+  }
+}
+
+func testFirstPerformance() {
+  // accessing `first` should not perform duplicate work on lazy collections
+  let col_ = (0..<10).lazy.filter({ print("filter: \($0)"); return (2..<8).contains($0) })
```

Comment:
>  it looks like `DefaultedForwardCollection` cannot, in fact, wrap any other collection

You're right.  Sorry, I didn't check when I was writing my comment.

> I also have no idea what this `_CollectionState` stuff is that the `Minimal*Index` types hold on to.

It checks index invalidation rules (see `docs/IndexInvalidation.rst`).

> I'm not sure what `Minimal*Collection` is for in the first place

It implements the protocol as narrowly as possible (for example, with regards to index invalidation and bounds checking).  Therefore, I don't think that allowing it to wrap another collection would make much sense.

> requiring existing clients to switch to `Defaulted*Collection<Minimal*Collection<T>>`

I think this would be the long-term strategy, but I don't think it would be palatable now when we don't have default generic parameters -- test code would become verbose everywhere.

> go back to having my own minimal collection wrapper and just use it to wrap `LoggingCollection`

I think that for now this would be the easiest approach to add this test.

Feel free to also add a test with a stateful `filter()` predicate (somewhere in `test/1_stdlib/Lazy.swift.gyb`), but I view it as an integration test for the lazy subsystem -- but since there are too many moving pieces here, I would like to have a direct test for the important behavior of `first` as you described above.